File: src/main/java/no/fint/provider/springer/storage/Wrapper.java
package no.fint.provider.springer.storage;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mongodb.BasicDBObject;
import org.jooq.lambda.Unchecked;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.stereotype.Service;

import java.util.function.Function;

@Service
public class Wrapper {

    @Autowired
    private ObjectMapper objectMapper;

    public Function<Object, Springer> wrapper(Class<?> type) {
        return Unchecked.function((Object content) -> {
            BasicDBObject basicDBObject = BasicDBObject.parse((objectMapper.writeValueAsString(content)));
            return new Springer(null, type.getCanonicalName(), basicDBObject);
        });
    }

    public <T> Function<Springer, T> unwrapper(Class<T> type) {
        return (Springer springer) -> objectMapper.convertValue(springer.getValue(), type);
    }

    public <T> Query query(Class<T> type) {
        return new Query().addCriteria(Criteria.where("type").is(type.getCanonicalName()));
    }

    public <T> Springer update(Springer springer, T content) {
        springer.setValue(BasicDBObject.parse(Unchecked.function(objectMapper::writeValueAsString).apply(content)));
        return springer;
    }
}

File: src/main/kotlin/no/fint/provider/springer/storage/Wrapper.kt
package no.fint.provider.springer.storage

import com.fasterxml.jackson.databind.ObjectMapper
import com.mongodb.BasicDBObject
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import org.springframework.stereotype.Service

@Service
class Wrapper(
    private val objectMapper: ObjectMapper
) {
    fun wrapper(type: Class<*>): (Any?) -> Springer = { content: Any? ->
        val basicDBObject = BasicDBObject.parse(objectMapper.writeValueAsString(content))
        Springer(null, type.canonicalName, basicDBObject)
    }

    fun <T> unwrapper(type: Class<T>): (Springer) -> T = { springer ->
        objectMapper.convertValue(springer.value, type)
    }

    fun <T> query(type: Class<T>): Query = Query().addCriteria(Criteria.where("type").`is`(type.canonicalName))

    fun <T> update(springer: Springer, content: T): Springer {
        springer.value = BasicDBObject.parse(objectMapper.writeValueAsString(content))
        return springer
    }
}
